### PR TITLE
add unit test for edge/pkg/devicetwin/dtmodule

### LINKED ISSUE #2240: edge/pkg/devicetwin/dtmodule: import test coverage
Add more tests to import test coverage. Details refer to #2110 .

/help

### PR DESCRIPTION
**Pull request by charles-woshicai**
**What type of PR is this?**
/kind test

**What this PR does / why we need it**:
Imporve unit test coverage of one go module.

**Which issue(s) this PR fixes**:
Ref #2240

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
None

```release-note

```

### CODE REVIEW

File: edge/pkg/devicetwin/dtmodule/dtmodule_test.go [hunk 1]
```diff
@@ -17,12 +17,17 @@ limitations under the License.
...
 	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcontext"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtmanager"
 	. "github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtmanager"
 	. "github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtmodule"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/stretchr/testify/assert"
+	"reflect"
+	"testing"
+	"time"
```
Comment (Iceber) [inline edge/pkg/devicetwin/dtmodule/dtmodule_test.go:30]: please format imports.
I don't think it's necessary to use `. "github.com/kubeedge/kubeedge/edge/pkg/devicetwin/<dtmanager, dtmanager>"`

Comment (charles-woshicai) [reply edge/pkg/devicetwin/dtmodule/dtmodule_test.go:30]: Will do.

**Comment by charles-woshicai** [comment]
> There are some errors with `make lint` in my local environment, anyone could help?
> 
> ```shell
> ➜  ~ echo $OSTYPE
> darwin20.0
> ➜  make-rules git:(test-coverage) /bin/zsh /Users/i535989/Dev/Projects/github.com/charles-woshicai/kubeedge/hack/make-rules/lint.sh
> /path/to/kubeedge/hack/make-rules/lint.sh:23: BASH_SOURCE[0]: parameter not set
> ```

Run `brew install gnu-sed` fix the error of `make lint`. Above error is not equal to the error of running `make lint`.

**Comment by charles-woshicai** [comment]
There is an error while running `keadm e2e-test` to  install the manifest of crd `devices.devices.kubeedge.io`.

Not sure how to fix this error of `keadm e2e-test`.

Below is the error
```shell
Error: CustomResourceDefinition.apiextensions.k8s.io "devices.devices.kubeedge.io" is invalid: spec.versions: Invalid value: []apiextensions.CustomResourceDefinitionVersion{apiextensions.CustomResourceDefinitionVersion{Name:"v1alpha2", Served:true, Storage:true, Deprecated:false, DeprecationWarning:(*string)(nil), Schema:(*apiextensions.CustomResourceValidation)(0xc00d5d2a40), Subresources:(*apiextensions.CustomResourceSubresources)(nil), AdditionalPrinterColumns:[]apiextensions.CustomResourceColumnDefinition(nil)}}: per-version schemas may not all be set to identical values (top-level validation should be used instead)
```

**Comment by fisherxu** [comment]
@charles-woshicai Please rebase to master branch to fix the failed ci.

**Comment by charles-woshicai** [comment]
Review needed.

File: edge/pkg/devicetwin/dtmodule/dtmodule_test.go [hunk 2]
```diff
@@ -75,3 +80,135 @@ var _ = Describe("Dtmodule", func() {
 		})
 	})
 })
+
+func TestDTModule_InitWorker(t *testing.T) {
```
Comment (Iceber) [inline edge/pkg/devicetwin/dtmodule/dtmodule_test.go:84]: TestDTModule_InitWorker and "Testing for InitWorker" context test look like duplicates

Comment (Iceber) [reply edge/pkg/devicetwin/dtmodule/dtmodule_test.go:84]: I think this BDD test can be removed, ginkgo might be a bit heavy for unit testing

Comment (charles-woshicai) [reply edge/pkg/devicetwin/dtmodule/dtmodule_test.go:84]: Yeah it's duplicates, I realize it now.

Will remove BDD test.

File: edge/pkg/devicetwin/dtmodule/dtmodule_test.go [hunk 3]
```diff
@@ -75,3 +80,135 @@ var _ = Describe("Dtmodule", func() {
...
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			dm := &DTModule{
+				Name: tt.fields.Name,
+			}
+			dm.InitWorker(recvCh, confirmCh, heartBearCh, ctx)
+			if !reflect.DeepEqual(dm.Worker, tt.want) {
```
Comment (Iceber) [inline edge/pkg/devicetwin/dtmodule/dtmodule_test.go:163]:
```go
assert.Equal(t, tt.want, dm.Worker)
```

File: edge/pkg/devicetwin/dtmodule/dtmodule_test.go [hunk 4]
```diff
@@ -75,3 +80,135 @@ var _ = Describe("Dtmodule", func() {
...
+			_, ok := ctx.ModulesHealth.Load(dtcommon.MemModule)
+			assert.False(t, ok)
+			go dm.Start()
+			ping := "ping"
+			tt.fields.HeartBeatCh <- ping
+			tt.fields.HeartBeatCh <- ping
+			lastTime, ok := ctx.ModulesHealth.Load(dtcommon.MemModule)
+			assert.True(t, ok)
+			healthTime, ok := lastTime.(int64)
+			assert.True(t, ok)
```
Comment (Iceber) [inline edge/pkg/devicetwin/dtmodule/dtmodule_test.go:210]: It looks better this way:
```go
if assert.True(t, ok) {
    t.Logf("last health time: %d", healthTime)
}
```

File: edge/pkg/devicetwin/dtmodule/dtmodule_test.go [hunk 5]
```diff
@@ -75,3 +80,135 @@ var _ = Describe("Dtmodule", func() {
...
+func TestDTModule_Start(t *testing.T) {
+	type fields struct {
+		Name        string
+		RecvCh      chan interface{}
+		ConfirmCh   chan interface{}
+		HeartBeatCh chan interface{}
+		Ctx         *dtcontext.DTContext
+	}
+	ctx, err := dtcontext.InitDTContext()
+	assert.Nil(t, err)
```
Comment (Iceber) [inline edge/pkg/devicetwin/dtmodule/dtmodule_test.go:179]: Actually using `require.Nil` is better, but the error returned by the `InitDTContext` is always nil, so `require` may not be necessary
```go
"github.com/stretchr/testify/require"
```

File: edge/pkg/devicetwin/dtmodule/dtmodule_test.go [hunk 6]
```diff
@@ -17,61 +17,143 @@ limitations under the License.
...
 
+	"github.com/stretchr/testify/assert"
+
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
 	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcontext"
 	. "github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtmanager"
 	. "github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtmodule"
 )
 
-var _ = Describe("Dtmodule", func() {
```
Comment (Iceber) [inline edge/pkg/devicetwin/dtmodule/dtmodule_test.go:28]: For removing the BDD test，`dtmodule_suite_test.go` also need to be removed.

**Comment by charles-woshicai** [comment]
Need a maintainer to approve running workflows,

**Comment by charles-woshicai** [comment]
@Iceber Could you approve me to run workflows?

**Comment by fisherxu** [comment]
> @Iceber Could you approve me to run workflows?

Done.

**Comment by charles-woshicai** [comment]
> > @Iceber Could you approve me to run workflows?
> 
> Done.

Thx :-D

**Review by fisherxu — COMMENTED** [review]
/approve
Thanks for the contribution :) @charles-woshicai
